v2/api_error.go

```diff
@@ -0,0 +1,185 @@
+type ErrDetails struct {
+	ErrorInfo           *errdetails.ErrorInfo
+	BadRequest          *errdetails.BadRequest
+	PreconditionFailure *errdetails.PreconditionFailure
+	QuotaFailure        *errdetails.QuotaFailure
+	RetryInfo           *errdetails.RetryInfo
+	ResourceInfo        *errdetails.ResourceInfo
+	RequestInfo         *errdetails.RequestInfo
+	DebugInfo           *errdetails.DebugInfo
+	Help                *errdetails.Help
+	LocalizedMessage    *errdetails.LocalizedMessage
+
+	//Unknown stores unidentifiable error details
+	Unknown []interface{}
+}
+
+type APIError struct {
+	err     error
+	status  *status.Status
+	details ErrDetails
+}
+
+//Details presents the error details in an APIError
+func (a *APIError) Details() ErrDetails {
+	return a.details
+}
+
+//Unwrap extracts original error
+func (a *APIError) Unwrap() error {
+	return a.err
+}
+
+//Error creates a readable representation of the APIError
+func (a *APIError) Error() string {
+	var d strings.Builder
+	d.WriteString("Error Details: ")
+
+	if a.details.ErrorInfo != nil {
+		d.WriteString("api error: name = ErrorInfo" + " reason = " + a.details.ErrorInfo.Reason +
+			" domain = " + a.details.ErrorInfo.Domain + "\n")
+	}
+
+	if a.details.BadRequest != nil {
+		v := a.details.BadRequest.GetFieldViolations()
+		var f []string
+		var desc []string
+		for _, x := range v {
+			f = append(f, x.Field)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = BadRequest" + " field = " + strings.Join(f, " ") +
+			" desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.PreconditionFailure != nil {
+		v := a.details.PreconditionFailure.GetViolations()
+		var t []string
+		var s []string
+		var desc []string
+		for _, x := range v {
+			t = append(t, x.Type)
+			s = append(s, x.Subject)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = PreconditionFailure" + " type = " + strings.Join(t, " ") +
+			" subj = " + strings.Join(s, " ") + " desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.QuotaFailure != nil {
+		v := a.details.QuotaFailure.GetViolations()
+		var s []string
+		var desc []string
+		for _, x := range v {
+			s = append(s, x.Subject)
+			desc = append(desc, x.Description)
+		}
+		d.WriteString("api error: name = QuotaFailure" + " subj = " + strings.Join(s, " ") +
+			" desc = " + strings.Join(desc, " ") + "\n")
+	}
+
+	if a.details.RequestInfo != nil {
+		d.WriteString("api error: name = RequestInfo" + " id = " + a.details.RequestInfo.RequestId +
+			" data = " + a.details.RequestInfo.ServingData + "\n")
+	}
+
+	if a.details.ResourceInfo != nil {
+		d.WriteString("api error: name = ResourceInfo" + " type = " + a.details.ResourceInfo.ResourceType +
+			" resourcename = " + a.details.ResourceInfo.ResourceName + " owner = " + a.details.ResourceInfo.Owner +
+			" desc = " + a.details.ResourceInfo.Description + "\n")
+
+	}
+	if a.details.RetryInfo != nil {
```

Comment:
I was wondering if it would be more useful to include some context for RetryInfo like "Please retry request in blank seconds" since RetryInfo only has seconds and nanos as its fields which I'm unsure on how useful it'll be to the user @noahdietz @summer-ji-eng

Reply:
Yeah that's a good idea. Keep it short and sweet though e.g. `retry in 5.014 seconds`. There is a way to convert the raw `google.protobuf.Duration` Go type into a Go `time.Duration` via https://pkg.go.dev/google.golang.org/protobuf/types/known/durationpb, which will make it easier to print the combined `seconds` and `nanos` in more human readable format.